citycar/generator.py

```diff
@@ -7,36 +7,32 @@
 
 import sys
 import random
+import argparse
 
 
-def help():
-    print("usage: generator.py <rows> <columns> <n_cars> <n_garage> <sparse> [<seed>]")
-    print("\t row and columns indicate the grid size. They should be > 1 ")
-    print("\t n_cars indicates how many cars have to go through the network")
-    print("\t n_garage indicates the number of starting garages")
-    print(
-        "\t if sparse != 0, the network has 'holes', i.e. junctions that cannot be used"
-    )
-    print("\t seed is the (optional) random seed for the instance")
-    sys.exit(2)
+to_print_car_pos = ""
 
+parser = argparse.ArgumentParser()
 
-to_print_car_pos = ""
+parser.add_argument("rows",type=int,help="the number of grid rows.")
+parser.add_argument("columns",type=int,help="the number of grid columns.")
+parser.add_argument("cars",type=int,help="how many cars have to go through the network")
+parser.add_argument("garages",type=int,help="the number of starting garages")
+parser.add_argument("--density",type=float,default=1.0,help="The ratio of the available roads.")
+parser.add_argument("--seed",type=int,help="random seed")
+
+args = parser.parse_args()
 
-if len(sys.argv) not in [6, 7]:
-    help()
-    sys.exit(2)
+row     = args.rows
+column  = args.columns
+car     = args.cars
+garage  = args.garages
+density = args.density
+seed    = args.seed
 
-row = int(sys.argv[1])
-column = int(sys.argv[2])
-car = int(sys.argv[3])
-garage = int(sys.argv[4])
-sparse = int(sys.argv[5])
+if density < 0.0 or 1.0 < density:
```

Comment:
We could use `if not (0.0 <= density <= 1.0):`

Reply:
oh well, what kind of method override trick python is using to allow it to be a valid code... I really hate python.
All languages should be like
> (<= 0.0 density 1.0)